File: jas/model_test.py
from absl.testing import absltest

from document import Document
from element import Layer, Rect
from model import Model


class ModelTest(absltest.TestCase):

    def test_default_document(self):
        model = Model()
        self.assertEqual(model.document.title, "Untitled")
        self.assertEqual(model.document.layers, ())

    def test_initial_document(self):
        doc = Document(title="Test")
        model = Model(document=doc)
        self.assertEqual(model.document.title, "Test")

    def test_set_document_notifies(self):
        model = Model()
        received = []
        model.on_document_changed(lambda doc: received.append(doc.title))
        model.document = Document(title="Changed")
        self.assertEqual(received, ["Changed"])

    def test_multiple_listeners(self):
        model = Model()
        a, b = [], []
        model.on_document_changed(lambda doc: a.append(doc.title))
        model.on_document_changed(lambda doc: b.append(doc.title))
        model.document = Document(title="X")
        self.assertEqual(a, ["X"])
        self.assertEqual(b, ["X"])

    def test_listener_called_on_each_change(self):
        model = Model()
        count = []
        model.on_document_changed(lambda doc: count.append(len(doc.layers)))
        layer = Layer(children=(), name="L1")
        model.document = Document(layers=(layer,))
        model.document = Document(layers=(layer, layer))
        self.assertEqual(count, [1, 2])

    def test_immutability(self):
        model = Model()
        before = model.document
        model.document = Document(title="New")
        after = model.document
        self.assertEqual(before.title, "Untitled")
        self.assertEqual(after.title, "New")


if __name__ == "__main__":
    absltest.main()

File: jas_ocaml/test/model_test.ml
let () =
  (* Test default document *)
  let model = Jas.Model.create () in
  assert (model#document.Jas.Document.title = "Untitled");
  assert (List.length model#document.Jas.Document.layers = 1);

  (* Test initial document *)
  let doc = Jas.Document.make_document ~title:"Test" [] in
  let model2 = Jas.Model.create ~document:doc () in
  assert (model2#document.Jas.Document.title = "Test");

  (* Test set_document notifies *)
  let model3 = Jas.Model.create () in
  let received = ref [] in
  model3#on_document_changed (fun doc -> received := doc.Jas.Document.title :: !received);
  model3#set_document (Jas.Document.make_document ~title:"Changed" []);
  assert (!received = ["Changed"]);

  (* Test multiple listeners *)
  let model4 = Jas.Model.create () in
  let a = ref [] in
  let b = ref [] in
  model4#on_document_changed (fun doc -> a := doc.Jas.Document.title :: !a);
  model4#on_document_changed (fun doc -> b := doc.Jas.Document.title :: !b);
  model4#set_document (Jas.Document.make_document ~title:"X" []);
  assert (!a = ["X"]);
  assert (!b = ["X"]);

  (* Test immutability *)
  let model5 = Jas.Model.create () in
  let before = model5#document in
  model5#set_document (Jas.Document.make_document ~title:"New" []);
  let after = model5#document in
  assert (before.Jas.Document.title = "Untitled");
  assert (after.Jas.Document.title = "New");

  Printf.printf "All model tests passed.\n"
